gulp-tasks/utils/publish-helpers.js

```diff
@@ -0,0 +1,179 @@
+const GROUPED_BUILD_FILES = 'grouped-build-files';
+
+const doesDirectoryExist = async (directoryPath) => {
+  let stats = null;
+  try {
+    stats = await fs.stat(directoryPath);
+  } catch (err) {
+    // NOOP
+  }
+  return stats !== null;
+};
+
+const getBuildPath = (tagName) => {
+  const tempReleasePath = path.join(
+    __dirname, '..', '..', constants.GENERATED_RELEASE_FILES_DIRNAME);
+  return path.join(tempReleasePath, tagName);
+};
+
+const downloadGitCommit = async (tagName, gitBranch) => {
+  if (!tagName) {
+    throw new Error(`You must provide a tagName to 'downloadGitCommit()`);
+  }
+
+  if (!gitBranch) {
+    throw new Error(`You must provide a gitBranch to 'downloadGitCommit()`);
+  }
+
+  const sourceCodePath = path.join(getBuildPath(tagName), SOURCE_CODE_DIR);
+
+  logHelper.log(oneLine`
+    Download Git Commit ${logHelper.highlight(gitBranch)}.
+  `);
+
+  const dirExists = await doesDirectoryExist(sourceCodePath);
+  if (!dirExists) {
+    logHelper.log(`    Clone Git repo for branch: '${gitBranch}'.`);
+
+    await spawn('git', [
+      'clone',
+      '--branch', gitBranch,
+      '--depth', '1',
+      `http://github.com/${constants.GITHUB_OWNER}/${constants.GITHUB_REPO}.git`,
+      sourceCodePath,
+    ]);
+  } else {
+    logHelper.log(`   Git repo for branch '${gitBranch}' is cloned already.`);
+  }
+
+  return sourceCodePath;
+};
+
+const buildGitCommit = async (tagName) => {
+  const sourceCodePath = path.join(getBuildPath(tagName), SOURCE_CODE_DIR);
+
+  logHelper.log(oneLine`
+    Building Commit
+    ${logHelper.highlight(path.relative(process.cwd(), sourceCodePath))}.
+  `);
+
+  await spawn('npm', ['install'], {
+    cwd: sourceCodePath,
+  });
+
+  await spawn('gulp', ['build'], {
+    cwd: sourceCodePath,
+  });
+};
+
+/*
+ * This function will create a directory with the same name as the
+ * .tar.gz file it generates. This way when the file is extracted
+ * the folder structure will be the same.
+ */
+const groupBuildFiles = async (tagName, gitBranch) => {
+  const groupedBuildFiles =
+    path.join(getBuildPath(tagName), GROUPED_BUILD_FILES);
+  const dirExists = await doesDirectoryExist(groupedBuildFiles);
+  logHelper.log(oneLine`
+    Group Build Files into
+    ${logHelper.highlight(path.relative(process.cwd(), groupedBuildFiles))}.
+  `);
+
+  if (!dirExists) {
+    await downloadGitCommit(tagName, gitBranch);
+    await buildGitCommit(tagName);
+
+    const sourceCodePath = path.join(getBuildPath(tagName), SOURCE_CODE_DIR);
+
+    const pattern = path.posix.join(
+      sourceCodePath, 'packages', '**',
+      constants.PACKAGE_BUILD_DIRNAME, constants.BROWSER_BUILD_DIRNAME,
+      '*.{js,map}');
+
+    // Copy files from the source code and move into the grouped build
+    // directory. In others, have a flat file structure of just the built files.
+    const filesToIncludeInBundle = glob.sync(pattern);
+    for (let fileToInclude of filesToIncludeInBundle) {
```

Comment:
Switch this to use `await Promise.all()` so that the copies [aren't sequential](https://developers.google.com/web/fundamentals/primers/async-functions#careful_avoid_going_too_sequential):

```js
const copyPromises = files.map((file) => fs.copy(file, path.join(groupedBuildFiles, path.basename(file)));
await Promise.all(copyPromises);
```

Reply:
I did this originally and hit some weird issues.